include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -0,0 +1,683 @@
+     *   // the tape and reuse this data.
+     *   const bool is_recording
+     *     = ad_helper.enable_record_sensitivities(tape_index);
+     *   if (is_recording == true)
+     *   {
+     *     // In this block on places the majority of the operations described
+     *     // in the previous code block. The set of operations that are
+     *     // conducted here therefore includes the following steps:
+     *     // - Register independent variables
+     *     // - Extract the sensitive equivalent of the independent variables
+     *     // - Use the sensitive variables to compute the dependent variables
+     *     // - Register the dependent variables with the helper class
+     *   }
+     *   else
+     *   {
+     *     // Load the existing tape to be reused
+     *     ad_helper.activate_recorded_tape(tape_no);
+     *
+     *     // Set the new values of the independent variables where the recorded
+     *     // dependent functions are to be evaluated (and differentiated
+     *     // around).
+     *     ad_helper.set_independent_variable(...);
+     *   }
+     *
+     *   // Compute derivatives of the dependent variables
+     *   const auto derivatives = ad_helper.compute_gradients();
+     * @endcode
+     *
+     * The second approach outlined here is more general than the first, and
+     * will work equally well for both taped and tapeless auto-differentiable
+     * numbers.
+     *
+     * @warning ADOL-C does not support the standard threading models used by
+     * deal.II, so this class should \b not be embedded within a multithreaded
+     * function when using ADOL-C number types. It is, however, suitable for use
+     * in both serial and MPI routines.
+     *
+     * @author Jean-Paul Pelteret, 2016, 2017, 2018
+     */
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType = double>
+    class ADHelperBase
+    {
+    public:
+      /**
+       * Type definition for the floating point number type that are used in,
+       * and result from, all computations.
+       */
+      using scalar_type =
+        typename AD::NumberTraits<ScalarType, ADNumberTypeCode>::scalar_type;
+
+      /**
+       * Type definition for the auto-differentiation number type that is used
+       * in all computations.
+       */
+      using ad_type =
+        typename AD::NumberTraits<ScalarType, ADNumberTypeCode>::ad_type;
+
+      /**
+       * @name Constructor / destructor
+       */
+      //@{
+
+      /**
+       * The constructor for the class.
+       *
+       * @param[in] n_independent_variables The number of independent variables
+       * that will be used in the definition of the functions that it is
+       * desired to compute the sensitivities of.
+       * @param[in] n_dependent_variables Number of scalar functions to be
+       * defined that will have a sensitivity to the given independent
+       * variables.
```

Comment:
is this a *subset of*, or *in addition to* the independent variables?

also, is it correct that these are the "sensitive" variables?

Reply:
I've now added some further text to clarify your misunderstanding here. Considering `\mathbf{f}(\mathbf{X})`, this is the number of outputs `\mathbf{f}`.